### DISCUSSION TITLE
Multiple addresses per invoice to retain Coinjoin UTXO privacy

### DISCUSSION THREAD

**Discussion by gunson21** [btcpayserver/btcpayserver#5447] [Ideas / Feature Requests]
When paying an on-chain invoice there is only one address for the whole sale.

However, if I have several post-mix UTXOs that are less than the invoice amount I would need to link them in order to pay the invoice. This would remove the privacy gained from the Coinjoin process.

What if a new address could be generated each time a UTXO is received? This way it would not be easy to link the UTXOs for other observers, although the merchant would know this (I only want to reveal this data to the merchant).

Or, the buyer could specify how many addresses they need so that they could quickly make all the transactions in parallel. Then the invoice is settled once all the transactions are confirmed.

**Comment by Kukks** [comment]
I share your concern; in fact, this was how we had it implemented for a long time until we removed it [here](https://github.com/btcpayserver/btcpayserver/pull/4984). Many merchants and their support staff were getting completely lost due to the last active address, and the payment addresses were different.  

One cool thing I am pushing forward is payments done through a coinjoin. This eliminates the need to make a tx that is dedicated to fulfilling a single payment (making it obvious that inputs being consolidated are from the same user). 

This comes with multiple benefits:
* No need to mix in advance, reducing mining fees
* Your change gets mixed for gains in anonymity.
* You can batch multiple payments
* You can add a surplus of coins for additional anonymity gains

The coinjoin plugin that is installable in BTCPay supports this already :)
